blameworthy/indexer.go

```diff
@@ -72,30 +72,84 @@ func (history GitHistory) FileBlame(commitHash string, path string) (*BlameResul
 	return &r, nil
 }
 
-func (history GitHistory) findCommit(commitHash string, path string) (File, int, error) {
+// Produces BlameVector for the given commits for the given path.
+// This method differs from FileBlame in that it does not try to compute previous and next commits.
+func (history GitHistory) FileBlameVectorBatch(commits []string, path string) ([]BlameVector, error) {
+        fileHistory, indices, err := history.FindCommitBatch(commits, path)
+	if err != nil {
+		return nil, err
+	}
+	resultMap := make(map[string]BlameVector)
+	for _, index := range indices {
+	    commitHash := fileHistory[index - 1].Commit.Hash
+	    resultMap[commitHash] = nil
+	}
+	segments := BlameSegments{}
+	commitsFound := 0
+	for _, diff := range fileHistory {
+	        commit := diff.Commit.Hash
+		segments = diff.step(segments)
+		if _, ok := resultMap[commit]; ok {
+		    resultMap[commit] = segments.flatten()
+		    commitsFound += 1
+		    if commitsFound == len(commits) {
+		            break
+		    }
+		}
+	}
+	resultList := make([]BlameVector, len(commits))
+	for i, index := range indices {
+	    commitHash := fileHistory[index - 1].Commit.Hash
+	    resultList[i] = resultMap[commitHash]
+	}
+	return resultList, nil
+}
+
+
+func (history GitHistory) FindCommitBatch(commitHashes []string, path string) (File, []int, error) {
 	fileHistory, ok := history.Files[path]
 	if !ok {
-		return File{}, -1, fmt.Errorf("no such file: %v", path)
+		return File{}, nil, fmt.Errorf("no such file: %v", path)
 	}
-	i := 0
+
+	commitMap := make(map[string]int)  // Maps a commit hash to its index into fileHistory
+	for _, commitHash := range commitHashes {
+	        commitMap[commitHash] = -1
+	}
+
 	j := 0
-	for ; i < len(history.Hashes); i++ {
+	commitsFound := 0
+	for i := 0; i < len(history.Hashes) && commitsFound < len(commitHashes); i++ {
 		h := history.Hashes[i]
 		if j < len(fileHistory) && fileHistory[j].Commit.Hash == h {
 			j++
 		}
-		if h == commitHash {
-			break
+                if _, ok := commitMap[h]; ok {
+                        commitMap[h] = j
+                        commitsFound += 1
+                }
+	}
+	indices := make([]int, len(commitHashes))
+	for i, commitHash := range commitHashes {
```

Comment:
The code is a little bit more general than it could be, but it does allow `commitHashes` to contain duplicates or be out of order.

You're correct that the error checking could be done by just iterating over `commitMap`. But in order to build `indices` to be the same length as the input `commitHashes`, I think you have to iterate over `commitHashes`.

Let me know if you feel we should just assert that `commitHashes` contains unique hashes that are ordered chronologically.